cirq-core/cirq/protocols/apply_unitary_protocol.py

```diff
@@ -454,7 +467,27 @@ def _strat_apply_unitary_from_decompose(val: Any, args: ApplyUnitaryArgs) -> Opt
     operations, qubits, _ = _try_decompose_into_operations_and_qubits(val)
     if operations is None:
         return NotImplemented
-    return apply_unitaries(operations, qubits, args, None)
+    all_qubits = frozenset([q for op in operations for q in op.qubits])
+    ancilla = tuple(sorted(all_qubits.difference(qubits)))
+    if not len(ancilla):
+        return apply_unitaries(operations, qubits, args, None)
+    ordered_qubits = ancilla + tuple(qubits)
+    all_qid_shapes = qid_shape_protocol.qid_shape(ordered_qubits)
+    state = qis.eye_tensor(all_qid_shapes, dtype=np.complex128)
+    buffer = np.empty_like(state)
```

Comment:
I think we should create a `kronecker_product` and `factor` method on `ApplyUnitaryArgs`; similar to `SimulationState`; and then dispatch the logic of allocate new space and extracting out the unitary corresponding to a subspace using function calls to these methods.